src/coroutine/memoset/mod.rs

```diff
@@ -178,69 +183,222 @@ impl<F: LurkField> CircuitTranscript<F> {
+                &self.result,
+                deps.hash(),
+            )
+        })
+    }
+}
+
+type PW<'a, F> = WithStore<'a, F, Provenance>;
+
+impl<'a, F: LurkField> Debug for PW<'a, F> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        let p = self.inner();
+        let s = self.store();
+
+        f.debug_struct("Provenance")
+            .field("full", &p.to_ptr(s).fmt_to_string_simple(s))
+            .field("query", &p.query.fmt_to_string_simple(s))
+            .field(
+                "dependencies",
+                &p.dependencies
+                    .iter()
+                    .map(|ptr| ptr.fmt_to_string_simple(s))
+                    .join(", "),
+            )
+            .finish()
+    }
+}
+
 #[derive(Clone, Debug)]
 /// A `Scope` tracks the queries made while evaluating, including the subqueries that result from evaluating other
 /// queries -- then makes use of the bookkeeping performed at evaluation time to synthesize proof of each query
 /// performed.
-pub struct Scope<Q, M> {
+pub struct Scope<Q, M, F: LurkField> {
     memoset: M,
     /// k => v
     queries: HashMap<Ptr, Ptr>,
     /// k => ordered subqueries
     dependencies: HashMap<Ptr, Vec<Q>>,
+    /// inverse of dependencies
+    dependents: HashMap<Ptr, HashSet<Ptr>>,
     /// kv pairs
     toplevel_insertions: Vec<Ptr>,
     /// internally-inserted keys
     internal_insertions: Vec<Ptr>,
     /// unique keys: query-index -> [key]
     unique_inserted_keys: HashMap<usize, Vec<Ptr>>,
-    transcribe_internal_insertions: bool,
     // This may become an explicit map or something allowing more fine-grained control.
+    provenances: OnceCell<HashMap<ZPtr<Tag, F>, ZPtr<Tag, F>>>,
     default_rc: usize,
 }
 
 const DEFAULT_RC_FOR_QUERY: usize = 1;
-const DEFAULT_TRANSCRIBE_INTERNAL_INSERTIONS: bool = false;
 
-impl<F: LurkField, Q> Default for Scope<Q, LogMemo<F>> {
+impl<F: LurkField, Q: Query<F>> Default for Scope<Q, LogMemo<F>, F> {
     fn default() -> Self {
-        Self::new(DEFAULT_TRANSCRIBE_INTERNAL_INSERTIONS, DEFAULT_RC_FOR_QUERY)
+        Self::new(DEFAULT_RC_FOR_QUERY)
     }
 }
 
-impl<F: LurkField, Q> Scope<Q, LogMemo<F>> {
-    fn new(transcribe_internal_insertions: bool, default_rc: usize) -> Self {
+impl<F: LurkField, Q: Query<F>> Scope<Q, LogMemo<F>, F> {
+    fn new(default_rc: usize) -> Self {
         Self {
             memoset: Default::default(),
             queries: Default::default(),
             dependencies: Default::default(),
+            dependents: Default::default(),
             toplevel_insertions: Default::default(),
             internal_insertions: Default::default(),
             unique_inserted_keys: Default::default(),
-            transcribe_internal_insertions,
+            provenances: Default::default(),
             default_rc,
         }
     }
+
+    fn provenance(&self, store: &Store<F>, query: Ptr) -> Result<Provenance, MemoSetError> {
```

Comment:
Would be nice to have a docstring for this function (and also for `provenance_from_kv`)